core/pta/attestation.c

```diff
@@ -0,0 +1,588 @@
+
+	if (att_data.pub)
+		free(att_data.pub);
+
+	if (att_data.enc_priv)
+		free(att_data.enc_priv);
+
+	if (att_data.sig)
+		free(att_data.sig);
+
+	return reason;
+}
+
+static TEE_Result priv_asn1_decode(uint8_t *priv, const size_t priv_len,
+				   uint8_t **d, size_t *d_size)
+{
+	uint32_t size = 0;
+	uint32_t offset = 7; // Beginn at the start of the modulus
+
+	if (priv_len < offset + 3)
+		return TEE_ERROR_SECURITY;
+
+	// Check the header
+	if (priv[0] != 0x30 ||
+	    priv[1] != 0x82 ||
+	    priv[offset] != 0x02 ||
+	    priv[offset + 1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	// Get the size of the first integer aka modulus
+	size = (priv[offset + 2] << 8) | priv[offset + 3];
+	offset += size + 4;
+
+	// Get the size of the second integer aka exponent
+	if (priv[offset] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	size = priv[offset + 1];
+	offset += size + 2;
+
+	if (priv_len < offset + 3)
+		return TEE_ERROR_SECURITY;
+
+	// Decoding d
+	if (priv[offset] != 0x02 || priv[offset + 1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (priv[offset + 2] << 8) | priv[offset + 3];
+
+	if (priv_len < offset + 4 + size)
+		return TEE_ERROR_SECURITY;
+
+	*d = &priv[offset + 4];
+	*d_size = size;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pub_asn1_decode(uint8_t **n, uint8_t **e,
+				  size_t *n_size, size_t *e_size)
+{
+	uint32_t size = 0;
+
+	// Decoding header
+	if (att_data.pub[0] != 0x30 || att_data.pub[1] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[2] << 8) | att_data.pub[3];
+	if ((size + 4) != att_data.pub_size)
+		return TEE_ERROR_SECURITY;
+
+	// Decoding first integer aka modulus
+	if (att_data.pub[4] != 0x02 || att_data.pub[5] != 0x82)
+		return TEE_ERROR_SECURITY;
+
+	size = (att_data.pub[6] << 8) | att_data.pub[7];
+	*n = &att_data.pub[8];
+	*n_size = size;
+
+	// Decoding second integer aka exponent
+	if (att_data.pub[size + 8] != 0x02)
+		return TEE_ERROR_SECURITY;
+
+	*e_size = att_data.pub[size + 9];
+	*e = &att_data.pub[size + 10];
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_sys_measurement(uint8_t *ptr)
+{
+	// TODO: Dummy function
+	// get_sys_measurement -> do something hardware specific
+	uint8_t sys[] = {0x00, 0x01, 0x02, 0x03,
+			0x04, 0x05, 0x06, 0x07,
+			0x08, 0x09, 0x0a, 0x0b,
+			0x0c, 0x0d, 0x0e, 0x0f,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x00, 0x00 };
+	memcpy(ptr, sys, 32);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_endorsement_key(uint8_t *key)
+{
+	// TODO: Dummy function get_endorsement_key
+	// do something hardware specific
+	uint8_t end_key[] = {0x9a, 0x04, 0xaa, 0x18,
+			0x2d, 0x03, 0x96, 0x74,
+			0x70, 0x8c, 0xe8, 0x07,
+			0xed, 0x91, 0x4c, 0xd1,
+			0x53, 0xcd, 0x9d, 0xf7,
+			0x80, 0x5e, 0x61, 0x74,
+			0x2f, 0x0a, 0xe4, 0x12,
+			0x94, 0x75, 0x8d, 0xd3 };
+	memcpy(key, end_key, 32);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result decrypt_priv_key(uint8_t *plain, size_t *plain_size)
```

Comment:
Thank you for reviewing my PR.

This function decrypts the attestation identifier key (AIK), which is stored in the attestation data struct (Image [here](https://github.com/linaro-swg/optee_examples/pull/76)). So this function does not decrypt the endorsement key this is done by the `get_endorsement_key` function. And [there ](https://github.com/OP-TEE/optee_os/pull/4011#discussion_r463608668)jfossier already requested the `__weak` symbol. Therefore I do not think the weak symbol is necessary here.

Reply:
Sorry, I thought about it again and after some research I understand your suggestion. You are absolutely right, I think it is the preferred solution to replace the `decrypt_priv_key()` function and ditch the `get_endorsementkey()` function. This will also keep the endorsement key in hardware.

Reply:
No issues, I think you understand now what I was trying to say. There are hardwares where endorsement key never leaves the SoC. In such cases, SoC vendors may want to override the complete decrypt_priv_key function as per their requirements.

Reply:
Fix [here](https://github.com/OP-TEE/optee_os/pull/4011/commits/5dd15cdf003c8ee35ffb269d92ef1b2ef64a5ad6)